### DISCUSSION TITLE
Support factory_args option for uvicorn.run() ?

### DISCUSSION THREAD

**Discussion by mottoslo** [Kludex/uvicorn#2377] [Ideas]
I am currently running uvicorn programmatically with single worker.
When running with single worker, application can be initialized and passed to uvicorn.run() function, therefore I could accept / apply application options from outside (command line ..etc)

However, as I proceed to change to using multiple workers, unvicorn.run() will only accept import string, not application instance itself.
as far as I know, config.load() will parse this string and import_from_string will retrieve application from this string.

For my app to be custom initialized with options provided from the command line, I would like to use the app_factory option. But as of now, self.loaded_app() (which is my app_factory method) is set to accept no argument.
https://github.com/encode/uvicorn/blob/c23cd24e6676ee0638d014d7000af1e6e0996bd6/uvicorn/config.py#L439-L440

I have made a workaround for app_factory to use environment variable instead, but I think Config should accept something like factory_args to be used by self.loaded_app() .

I am up for opening a PR for this matter if this gets confirmed as an issue !
Also, this is my first time raising an issue, so any feedback would be much appreciated :)

p.s) I did find a similar discussion page (https://github.com/encode/uvicorn/discussions/1644) but I don't think the link in the comment correctly addresses this problem

**Comment by SaravananSathyanandhaQC** [comment]
Second this, and have a very basic PR to handle this by supporting an arguments of `worker_kwargs=...` in `uvicorn.run`.

https://github.com/SaravananSathyanandhaQC/uvicorn/tree/worker-kwargs
https://github.com/encode/uvicorn/compare/master...SaravananSathyanandhaQC:uvicorn:worker-kwargs?expand=1

I have two use cases where this would be valuable (both in scenarios with multiple workers hence have to use the factory):

1. This is started via a CLI that passes some args, and some logic is done before starting the uvicorn workers, and I need to pass that information through. Currently the best way to do this is a whole load of `os.environ["key"] = value` for anything I want to pass through, but it's a bit verbose and only really works for strings, anything nested gets messy fast.

2. If I want to use shared memory across the workers (e.g. caching, shared state), this actually works very nicely with my branch above, it's very easy to initialise a `SyncManager.dict` and pass it down to the workers so they all can share the same dict.
